query/src/main/scala/filodb/query/exec/AggrOverRangeVectors.scala

```diff
@@ -164,6 +171,54 @@ object RangeVectorAggregator {
       }
     }
   }
+
+  /**
+   * A fast reduce method intended specifically for the case when no grouping needs to be done AND
+   * the previous transformer is a PeriodicSampleMapper with fixed output lengths.
+   * It's much faster than mapReduce() since it iterates through each vector first and then from vector to vector.
+   * Time wise first iteration also uses less memory for high-cardinality use cases and reduces the
+   * time window of holding chunk map locks to each time series, instead of the entire query.
+   */
+  def fastReduce(rowAgg: RowAggregator,
+                 skipMapPhase: Boolean,
+                 source: Observable[RangeVector],
+                 outputLen: Int): Observable[RangeVector] = {
+    // Can't use an Array here because rowAgg.AggHolderType does not have a ClassTag
+    val accs = collection.mutable.ArrayBuffer.fill(outputLen)(rowAgg.zero)
```

Comment:
Memory used is O(steps in the query).
What if use requests a long time duration. Shouldnt this optimization be applied only if small time ranges are chosen?

Reply:
Added a configurable max # of windows when the fastReduce path is chosen instead of the normal one.